src/main/java/uk/gov/companieshouse/search/api/service/upsert/advanced/AdvancedUpsertRequestService.java

```diff
@@ -0,0 +1,79 @@
+package uk.gov.companieshouse.search.api.service.upsert.advanced;
+
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.action.update.UpdateRequest;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import uk.gov.companieshouse.api.model.company.CompanyProfileApi;
+import uk.gov.companieshouse.environment.EnvironmentReader;
+import uk.gov.companieshouse.search.api.elasticsearch.AdvancedSearchUpsertRequest;
+import uk.gov.companieshouse.search.api.exception.UpsertException;
+import uk.gov.companieshouse.search.api.logging.LoggingUtils;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+@Service
+public class AdvancedUpsertRequestService {
+
+    @Autowired
+    private EnvironmentReader environmentReader;
+
+    @Autowired
+    private AdvancedSearchUpsertRequest advancedSearchUpsertRequest;
+
+    private static final String INDEX = "ADVANCED_SEARCH_INDEX";
+
+    /**
+     * Create an index request for document if it does not currently exist
+     * @param company - Company sent over in REST call to be added/updated
+     * @return {@link IndexRequest}
+     * @throws UpsertException
+     */
+    public IndexRequest createIndexRequest(CompanyProfileApi company, String orderedAlphaKey,
+                                           String sameAsKey) throws UpsertException {
+
+        Map<String, Object> logMap = setUpUpsertLogging(company);
+        try {
+            LoggingUtils.getLogger().info("Preparing index request", logMap);
+            return new IndexRequest(environmentReader.getMandatoryString(INDEX))
+                .source(advancedSearchUpsertRequest.buildRequest(company, orderedAlphaKey, sameAsKey)).id(company.getCompanyNumber());
+        } catch (IOException e) {
+            LoggingUtils.getLogger().error("Failed to index a document for company", logMap);
+            throw new UpsertException("Unable create index request");
+        }
+    }
+
+    /**
+     * If document already exists attempt to upsert the document
+     * @param company - Company sent over in REST call to be added/updated
+     * @param indexRequest
+     * @return {@link UpdateRequest}
+     * @throws UpsertException
+     */
+    public UpdateRequest createUpdateRequest(CompanyProfileApi company, String orderedAlphaKey,
+                                             String sameAsKey, IndexRequest indexRequest) throws UpsertException {
+
+        Map<String, Object> logMap = setUpUpsertLogging(company);
+        try {
+            LoggingUtils.getLogger().info("Attempt to upsert document if it does not exist", logMap);
+
+            return new UpdateRequest(environmentReader.getMandatoryString(INDEX), company.getCompanyNumber())
+                .docAsUpsert(true)
```

Comment:
Doesn't this line do the insert if it doesn't exist or update if it does? Is the insert endpoint needed if this is the case?

Reply:
Yeah I think you are right - will have a play around with that to see what is and isn't required to do an upsert.